### ISSUE TITLE
Failed lookup() short-circuits interpolation

### ISSUE THREAD

**Issue by daveadams** [hashicorp/terraform#4474]
Consider this terraform code:

    variable "food" {}

    variable "type_lookup" {
        default = {
            apple = "fruit"
            banana = "fruit"
            tomato = "vegetable"
            beef = "meat"
            spam = ""
        }
    }

    output "food_type" {
        value = "${coalesce(lookup(var.type_lookup, var.food), "unknown")}"
    }

So, given an input variable `food`, I would expect the module provide an output `food_type` containing the lookup value from the `type_lookup` map, or to return `unknown` if the food is not specified in the map (or if the specified value is an empty string, as with the key `spam` in the example).

However, what happens is that if the key provided to the lookup function is not found, the output `food_type` is just not returned at all.

Examples:

    $ terraform apply
    var.food
      Enter a value: spam
    
    Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
    
    Outputs:
      food_type = unknown
    
    $ terraform apply
    var.food
      Enter a value: apple
    
    Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
    
    Outputs:
      food_type = fruit
    
    $ terraform apply
    var.food
      Enter a value: durian
    
    Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

    [no outputs]

In other words, a failed `lookup()` call appears to short-circuit the entire interpolation. This is not what I would expect. Further, even reordering the arguments to `coalesce`, or adding fixed text on either side of the interpolation, or even just using the `lookup()` call alone also fails silently in the same way. So eg, this additional output:

    output "second_try" { value = "hello: ${lookup(var.type_lookup, var.food)}" }

This results in no output value for `second_try` if the value of `var.food` is not a key in the `var.type_lookup` map.

IMO, `lookup()` should either return an empty string on a failed lookup or provide a way to specify some default value. It definitely should not short-circuit the entire output block.

**Comment by phinze** [comment]
Thanks for the clear explanation here. Totally agreed that we should have a way of accomplishing this. I'm guessing that the easiest way to handle this would be to add a third "default" argument to `lookup()`.

**Comment by roderickrandolph** [comment]
:+1: I also came across this issue while attempting to concatenate a string of comma delimited `lookup()` data that I eventually want to `split()` and create a `compact()` array (e.g. to form a list of security group IDs). When the `lookup()` fails terraform essentially creates an empty array which can cause an unintentionally change/modification to the infrastructure's state (e.g. removing all security groups). So, in addition to adding a "default" argument option we may also want to set the "default"'s default to an empty string to minimize unintentional surprises.

**Comment by daveadams** [comment]
PR #6884 addresses this problem with an optional third argument to lookup(). I did not change the behavior of lookup() with two arguments, even though I don't think it's behaving quite as designed. (I think it is supposed to throw an error rather than silently failing and breaking outputs, etc, but lookup() is treated differently than other interpolation functions.)
